File: tests/math/tool4_coset.py
"""
Tool 4 — Coset Factorization Calculus (SKELETON).

Theorem 4.2 (hiding characterization) is a knowledge-state theorem: given
which subset of {π, ρ_1, ρ_2, s} a party knows, its posterior on π is one
of:
  - if knows ρ_2 but neither ρ_1 nor s: guessing prob ≤ γ(D)
  - if knows ρ_1 (or s) but not ρ_2:      posterior = prior (perfect hiding)

This file simulates the knowledge-state machinery empirically at small m by
enumerating the derivation closure over F = {π, ρ_1, ρ_2, s} and computing
each party role's posterior on π.

Verification target for this session (partial):
  - Enumeration of all "knowledge states" and their closure under
    derivation rules.
  - For each state, compute the posterior distribution of π given that
    state and check against Theorem 4.2's claimed classification.

Corollary 4.3 (coalition matrix) skeleton also below: for each pair of
parties, compute their JOINT knowledge state and derive coalition capabilities.

Fully implementing Theorem 4.2 requires:
  - A model of π (target injection [c] → [m]) — small m, small c makes this
    trivially enumerable.
  - ρ_1 ← D (from Tool 2).
  - ρ_2 = π · ρ_1^{-1} — derivable computation.

DEFERRED to next session:
  - The both-blind XOR-shared variant proof (Prop 1.2 iii applied to ρ_1).
  - Formal simulator specification for the UC treatment.
"""

import math
from itertools import combinations


# ---------------------------------------------------------------------------
# Knowledge-state closure
# ---------------------------------------------------------------------------

FACTORS = ["pi", "rho1", "rho2", "s"]

# Derivation rules: (subset of factors known → factor derived).
# From the theory doc:
#   {pi, rho1}    → rho2  (since rho2 = pi * rho1^-1)
#   {pi, rho2}    → rho1
#   {rho1, rho2}  → pi
#   {s}           → rho1
DERIVATION_RULES = [
    ({"pi", "rho1"}, "rho2"),
    ({"pi", "rho2"}, "rho1"),
    ({"rho1", "rho2"}, "pi"),
    ({"s"}, "rho1"),
]


def closure(initial_set):
    """Repeatedly apply derivation rules until no new factor added."""
    known = set(initial_set)
    changed = True
    while changed:
        changed = False
        for prereqs, out in DERIVATION_RULES:
            if prereqs.issubset(known) and out not in known:
                known.add(out)
                changed = True
    return known


def hiding_classification(state):
    """
    Given a closed knowledge state, classify per Theorem 4.2:
      "full"      — state contains pi (or can derive it) → no hiding
      "gamma_D"   — state has rho2 but neither rho1 nor s → guessing bounded by γ(D)
      "perfect"   — state has rho1 (or s) but not rho2 → posterior = prior
      "trivial"   — state has neither rho1 nor rho2 → posterior = prior (also perfect)
    """
    closed = closure(state)
    if "pi" in closed:
        return "full_leak"
    if "rho2" in closed and "rho1" not in closed and "s" not in closed:
        return "gamma_D"
    if ("rho1" in closed or "s" in closed) and "rho2" not in closed:
        return "perfect"
    return "trivial"


def print_all_knowledge_states():
    """Enumerate all 2^4 = 16 knowledge states, show closure + classification."""
    print(f"{'state':<28} {'closure':<38} {'classification'}")
    print("-" * 90)
    for k in range(5):
        for state in combinations(FACTORS, k):
            state_set = set(state)
            closed = closure(state_set)
            cls = hiding_classification(state_set)
            state_str = "{" + ", ".join(sorted(state_set)) + "}" if state_set else "{}"
            closed_str = "{" + ", ".join(sorted(closed)) + "}"
            print(f"{state_str:<28} {closed_str:<38} {cls}")


# ---------------------------------------------------------------------------
# Party role assignments per the theory doc (§4, "The split, derived rather
# than designed"): each party's initial knowledge state.
# ---------------------------------------------------------------------------

PARTY_ROLES = {
    "Router":      {"s"},         # knows s hence rho1; never receives rho2
    "Matcher":     {"pi", "rho2"}, # holds pi and derived rho2 (matcher role)
    "Party P_i":   {"rho2"},       # receives rho2 from matcher only
}


def show_party_classifications():
    print("\n--- Per-role hiding classification (Theorem 4.2) ---")
    for role, state in PARTY_ROLES.items():
        cls = hiding_classification(state)
        closed = closure(state)
        print(f"  {role:<15} initial={sorted(state)}  closure={sorted(closed)}  → {cls}")


def show_coalitions():
    """Corollary 4.3 — coalition matrix by union of knowledge states."""
    print("\n--- Coalition matrix (Corollary 4.3) ---")
    roles = list(PARTY_ROLES.keys())
    for i, r1 in enumerate(roles):
        for r2 in roles[i:]:
            joined = PARTY_ROLES[r1] | PARTY_ROLES[r2]
            cls = hiding_classification(joined)
            closed = closure(joined)
            print(f"  {r1} + {r2:<15} joined={sorted(joined)}  "
                  f"closure={sorted(closed)}  → {cls}")


def main():
    print("=== Tool 4 — Coset factorization (SKELETON) ===\n")
    print("Enumeration of all 2^4 = 16 knowledge states, with closure and")
    print("Theorem-4.2 hiding classification.\n")
    print_all_knowledge_states()
    show_party_classifications()
    show_coalitions()
    print("\nDEFERRED:")
    print("  - Explicit simulator computations of posterior distributions (needs")
    print("    Tool 2 D-distribution + small-m π enumeration; feasible next session).")
    print("  - Both-blind XOR-shared-control variant (Prop 1.2(iii) applied to ρ_1).")
    print("  - Formal UC composition of the coset factorization.")


if __name__ == "__main__":
    main()

File: tests/math_cpp/tool4_coset.cpp
// Tool 4 — Coset factorization: knowledge-state closure (C++ port).

#include <cstdio>
#include <set>
#include <string>
#include <vector>
#include <algorithm>

namespace {

using State = std::set<std::string>;
const std::vector<std::string> FACTORS = {"pi", "rho1", "rho2", "s"};

// Derivation rules: {prereq set} -> derived factor
struct Rule { std::set<std::string> prereq; std::string out; };
const std::vector<Rule> RULES = {
    {{"pi", "rho1"}, "rho2"},
    {{"pi", "rho2"}, "rho1"},
    {{"rho1", "rho2"}, "pi"},
    {{"s"}, "rho1"},
};

State closure(State known) {
    bool changed = true;
    while (changed) {
        changed = false;
        for (const auto& r : RULES) {
            if (std::includes(known.begin(), known.end(),
                              r.prereq.begin(), r.prereq.end())
                && known.find(r.out) == known.end()) {
                known.insert(r.out);
                changed = true;
            }
        }
    }
    return known;
}

std::string classify(const State& s0) {
    auto c = closure(s0);
    if (c.count("pi")) return "full_leak";
    if (c.count("rho2") && !c.count("rho1") && !c.count("s")) return "gamma_D";
    if ((c.count("rho1") || c.count("s")) && !c.count("rho2")) return "perfect";
    return "trivial";
}

std::string set_str(const State& s) {
    std::string r = "{";
    bool first = true;
    for (const auto& e : s) { if (!first) r += ","; r += e; first = false; }
    return r + "}";
}

void enumerate_all_states() {
    std::printf("%-28s %-38s %s\n", "state", "closure", "class");
    std::printf("%s\n", std::string(90, '-').c_str());
    int n = FACTORS.size();
    for (int mask = 0; mask < (1 << n); ++mask) {
        State s;
        for (int i = 0; i < n; ++i) if ((mask >> i) & 1) s.insert(FACTORS[i]);
        std::printf("%-28s %-38s %s\n",
                    set_str(s).c_str(), set_str(closure(s)).c_str(),
                    classify(s).c_str());
    }
}

const std::vector<std::pair<std::string, State>> ROLES = {
    {"Router",     {"s"}},
    {"Matcher",    {"pi", "rho2"}},
    {"Party P_i",  {"rho2"}},
};

}  // namespace

int main() {
    std::puts("=== Tool 4 — Coset factorization: knowledge closure (C++) ===\n");
    enumerate_all_states();

    std::puts("\n--- Per-role hiding classification (Theorem 4.2) ---");
    for (const auto& [role, s] : ROLES) {
        std::printf("  %-15s initial=%s  closure=%s  -> %s\n",
                    role.c_str(), set_str(s).c_str(),
                    set_str(closure(s)).c_str(), classify(s).c_str());
    }

    std::puts("\n--- Coalition matrix (Corollary 4.3) ---");
    for (size_t i = 0; i < ROLES.size(); ++i) {
        for (size_t j = i; j < ROLES.size(); ++j) {
            State joined = ROLES[i].second;
            for (const auto& x : ROLES[j].second) joined.insert(x);
            std::printf("  %s + %s: joined=%s -> %s\n",
                        ROLES[i].first.c_str(), ROLES[j].first.c_str(),
                        set_str(joined).c_str(), classify(joined).c_str());
        }
    }
    return 0;
}
